c/producer_seq: accept batches with last seq < than first seq

mmaslankaprv opened the pull request:
Batch identity with `fist_seq` > `last_seq` is a valid batch if an overflow in sequence number occurred. In this case a first sequence will be a large number (close to `int32_t` max value) while the last sequence will be wrapped around and difference will be added to `0`.

This is a valid situation and producer should still be able to continue. Returning an `invalid_request` error lead to a situation in which producer can not continue and partition leader is stepping down as an error is interpreted as an replication error by the `rm_stm`

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
### Bug Fixes
- Fixes handling batch sequence number overflow

mmaslankaprv committed:
c/producer_seq: accept batches with last seq < than first seq

Batch identity with `fist_seq` > `last_seq` is a valid batch if an
overflow in sequence number occurred. In this case a first sequence will
be a large number (close to `int32_t` max value) while the last sequence
will be wrapped around and difference will be added to `0`.

This is a valid situation and producer should still be able to continue.
Returning an `invalid_request` error lead to a situation in which
producer can not continue and partition leader is stepping down as an
error is interpreted as an replication error by the `rm_stm`

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/producer_state: simplified sequence number validation

The validation logic is very simple, if there is no previous state
incoming batch sequence number must be equal to zero. Otherwise it must
be equal to the next one to the last stored

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
c/tests: added test validating batch sequence number overflow

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/cluster/producer_state.cc:
```diff
@@ -132,15 +132,25 @@ void requests::reset(request_result_t::error_type error) {
+        // no previous request, expect incoming sequence to be 0
+        return incoming == 0;
+    }
+    return incoming == next_sequence(last_req.value()->_last_sequence);
```
Not sure if a negative value can make its way here, but if it can it's not only a simplification but a correction as well: previously incoming = INT_MIN would be accepted if `last_req.value()->_last_sequence` was INT_MAX. I can imagine it may be a common mistake since the overflow logic differs from natural. Could you update the commit comment accordingly and add a test?

mmaslankaprv replied:
no, it is always wrapped around 0 and positive

bashtanov replied:
What guarantees it's non-negative?

mmaslankaprv replied:
handler logic

mmaslankaprv replied:
I will think about a test maybe and add a comment

mmaslankaprv replied:
This is the place where it is computed
https://github.com/redpanda-data/redpanda/blob/1492a874cc7b464b1a9661745ffc8f9a8b20e16e/src/v/model/record.h#L650

mmaslankaprv replied:
I've added explicit handling, you were right. Some tests were using negative values which isn't correct.

bashtanov commented on src/v/cluster/tests/producer_state_tests.cc:
```diff
@@ -346,3 +346,81 @@ FIXTURE_TEST(test_state_management_with_multiple_namespaces, test_fixture) {
+        std::numeric_limits<seq_t>::max() - 15,
+        std::numeric_limits<seq_t>::max() - 10),
+      model::term_id(1));
+    //
```
nit: Formatting is a bit confusing to me. Could you maybe make each case (starting with a `try_emplace_request` line and ending with checks) run without blank lines, and separate cases with blank lines?

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cd9796ae-5576-40fa-b3ea-1163f3abc981-->
<!--TEST RESULTS-->
<!--9942f96d-b5c2-424f-b11f-0e760851c568-->
<!--TEST RESULTS-->
<!--1d046403-bfc5-4c97-acef-1cf575fcccec-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73335">build#73335</a></summary>


| test_class               | test_method             | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                   |
|:-------------------------|:------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodePostRestartProbeTest | post_restart_probe_test | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73335#01999ff7-77b9-4380-98b5-646d14aa3345 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodePostRestartProbeTest&test_method=post_restart_probe_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73405">build#73405</a></summary>


| test_class                               | test_method                              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   | test_history   |
|:-----------------------------------------|:-----------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|:---------------|
| src/v/cluster/tests/producer_state_tests | src/v/cluster/tests/producer_state_tests |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/73405#0199a340-408d-4ef5-b7a4-48faeca2ecbf | FAIL          | 0/1      |          |                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73439">build#73439</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    | test_history                                                                                                                                                             |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73439#0199a464-bded-417d-b081-e2d1b450e355 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| OffsetRetentionTest   | test_offset_expiration               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73439#0199a464-bdea-4c8c-8300-5d8711e5bfd4 | FAIL          | 0/1      |                                                                                                                                                           | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=OffsetRetentionTest&test_method=test_offset_expiration                 |
</details>
<!--/body-->

mmaslankaprv committed:
c/producer_state: explicitly handle negative batch sequence numbers

Batch sequence can never be negative when producer is idempotent. It is
required to start from '0' and it has a special handling for the
overflow. Added explicit handling of negative batch sequence numbers as
the other parts of producer state logic depends on the sequence being a
number greater than or equal to `0`

Signed-off-by: Michał Maślanka <michal@redpanda.com>

bashtanov commented on src/v/cluster/tests/producer_state_tests.cc:
```diff
@@ -347,29 +351,45 @@ FIXTURE_TEST(test_state_management_with_multiple_namespaces, test_fixture) {
+    if (first < 0 || last < 0) {
+        return 0;
+    }
+    if (last == first) {
+        return 1;
+    }
+    if (last > first) {
+        return last - first + 1;
```
nit: can be simplified as `if (last >= first) {return last - first + 1;}`

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73439-->
<!--9c9a69a4-558a-4d01-adcd-2d8e430e9bd5-->
<!--title-->
## Retry command for [Build#73439](https://buildkite.com/redpanda/redpanda/builds/73439)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/offset_retention_test.py::OffsetRetentionTest.test_offset_expiration
```
<!--/body-->

mmaslankaprv commented:
Unrelated failure fixed by: https://github.com/redpanda-data/redpanda/pull/27848

vbotbuildovich commented:
/backport v25.2.x

vbotbuildovich commented:
/backport v25.1.x

vbotbuildovich commented:
/backport v24.3.x

dotnwat commented on src/v/cluster/producer_state.cc:
```diff
@@ -351,10 +361,22 @@ void producer_state::reset_with_new_epoch(model::producer_epoch new_epoch) {
+    if (bid.first_seq < 0 || bid.last_seq < 0) {
+        vlog(
+          _logger.warn,
+          "[{}] request with non idempotent batch {}, term: {}, reset: {}",
+          *this,
+          bid,
+          current_term,
+          reset,
+          _requests);
```
@mmaslankaprv mismatched number of arguments and  `{}` placeholders?

bharathv replied:
@dotnwat https://github.com/redpanda-data/redpanda/pull/27921/

piyushredpanda replied:
Thank you, @dotnwat and @bharathv
